homeassistant/components/calendar/__init__.py

```diff
@@ -0,0 +1,286 @@
+        start = _get_date(event['start'])
+        end = _get_date(event['end'])
+
+        # check if the start/end times changed, if not then exit
+        if start == self._cal_data['start'] and end == self._cal_data['end']:
+            # don't redo the listener if we have the same start/end times
+            return
+
+        summary = event['summary']
+
+        # check if we have an offset tag in the message
+        # time is HH:MM or MM
+        reg = '{}([+-]?[0-9]{{0,2}}(:[0-9]{{0,2}})?)'.format(self._offset)
+        search = re.search(reg, summary)
+        if search and search.group(1):
+            time = search.group(1)
+            if ':' not in time:
+                if time[0] == '+' or time[0] == '-':
+                    time = '{}0:{}'.format(time[0], time[1:])
+                else:
+                    time = '0:{}'.format(time)
+
+            self._cal_data['offset_time'] = time_period_str(time)
+            summary = (summary[:search.start()] + summary[search.end():]) \
+                .strip()
+        else:
+            self._cal_data['offset_time'] = dt.dt.timedelta()  # default it
+
+        # Remove the search term if we're using tags
+        if self.search and self.search[0] == '#':
+            summary = re.compile(re.escape(self.search),
+                                 re.IGNORECASE).sub('', summary)
+
+        # cleanup the string so we don't have a bunch of double+ spaces
+        self._cal_data['message'] = re.sub('  +', '', summary).strip()
+
+        self._cal_data['location'] = event.get('location', '')
+        self._cal_data['description'] = event.get('description', '')
+        self._cal_data['start'] = start
+        self._cal_data['end'] = end
+
+        if self._cal_data['start'] < dt.now():
+            if not self.is_on:
+                # trigger the device open
+                self._state = True
+        elif self._state:
+            # we shouldn't be on right now
+            self._state = False
+
+        if not self._state:
+            # schedule the device to open
+            def _start(now):
+                """What to do when the event start time is reached."""
+                self._state = True
+                self._start_listener = None
+            if self._start_listener:
+                self.hass.bus.remove_listener(
+                    EVENT_TIME_CHANGED, self._start_listener)
+            self._start_listener = track_point_in_time(self.hass,
+                                                       _start,
+                                                       self._cal_data['start'])
+
+        # schedule the close
+        def _end(now):
+            """What to do when the event end time is reached."""
+            self._state = False
+            self.cleanup()
+        self._end_listener = track_point_in_time(self.hass, _end,
+                                                 self._cal_data['end'])
+
+    def value_changed(self, state):
```

Comment:
It appeared, months ago, that this was called by another process and I didn't have to deal with changing the state. I take it either I'm wrong about that and this was never needed?

Do I need to call update_ha_state when state data changes?